src/test/java/redis/clients/jedis/tests/SSLJedisClusterTest.java

```diff
@@ -0,0 +1,262 @@
+      // initial connection to localhost works, but subsequent connections to nodes use 127.0.0.1
+      // and fail hostname verification
+      Assert.assertEquals("The JedisConnectionException does not contain the expected message.",
+          "Could not get a resource from the pool", e.getMessage());
+    }
+    jc.close();
+  }
+  
+  @Test
+  public void connectByIpAddressWithSSLParameters() {
+    final SSLParameters sslParameters = new SSLParameters();
+    sslParameters.setEndpointIdentificationAlgorithm("HTTPS");
+    
+    JedisCluster jc = null;
+    try {
+      jc = new JedisCluster(new HostAndPort("127.0.0.1", 8379), DEFAULT_TIMEOUT, DEFAULT_TIMEOUT,
+          DEFAULT_REDIRECTIONS, "cluster", null, DEFAULT_CONFIG, true, 
+          null, sslParameters, null, portMap);
+      Assert.fail("The code did not throw the expected JedisConnectionException.");
+    } catch (JedisConnectionException e) {
+      Assert.assertEquals(SSLException.class, e.getCause().getClass());
+      Assert.assertEquals(SSLHandshakeException.class, e.getCause().getCause().getClass());
+      Assert.assertEquals(CertificateException.class, e.getCause().getCause().getCause().getClass());
+    }
+    
+    if (jc != null) {
+      jc.close();
+      Assert.fail("JedisCluster object is unexpectedly non-null");
+    }
+  }
+  
+  @Test
+  public void connectWithCustomHostNameVerifier() {
+    HostnameVerifier hostnameVerifier = new BasicHostnameVerifier();
+    HostnameVerifier localhostVerifier = new LocalhostVerifier();
+    
+    JedisCluster jc = new JedisCluster(new HostAndPort("localhost", 8379), DEFAULT_TIMEOUT, DEFAULT_TIMEOUT,
+        DEFAULT_REDIRECTIONS, "cluster", null, DEFAULT_CONFIG, true, 
+        null, null, hostnameVerifier, portMap);;
+    try {
+      // initial connection made with 'localhost' but subsequent connections to nodes use 127.0.0.1
+      // which causes custom hostname verification to fail
+      jc.get("foo");
+      Assert.fail("The code did not throw the expected JedisConnectionException.");
+    } catch (JedisConnectionException e) {
+      Assert.assertEquals("The JedisConnectionException does not contain the expected message.",
+          "Could not get a resource from the pool", e.getMessage());
+    }
+    jc.close();
+
+    JedisCluster jc2 = null;
+    try {
+      jc2 = new JedisCluster(new HostAndPort("127.0.0.1", 8379), DEFAULT_TIMEOUT, DEFAULT_TIMEOUT,
+          DEFAULT_REDIRECTIONS, "cluster", null, DEFAULT_CONFIG, true, 
+          null, null, hostnameVerifier, portMap);
+      Assert.fail("The code did not throw the expected NullPointerException.");
+    } catch (NullPointerException e) {
+      // Null pointer exception occurs from closing Jedis object that did not connect due to custom
+      // hostname validation. When closing this Jedis object, the RedisOutputStream in the underlying 
+      // Connection object is null and causes this exception
+    }
+    if (jc2 != null) {
```

Comment:
If `jc2 != null`, execution will not reach this point because of previous `Assert.fail()`. Put this block in finally and remove following `Assert.fail()`.